syntax = "proto3";

package quintet.rpc;

message PbReply {
  uint64 term = 1;
  bool ans = 2;
}

message PbServerId {
  string addr = 1;
  uint32 port = 2;
}

message PbLogEntry {
  uint64 term = 1;
  string opName = 2;
  string args = 3;
  uint64 prmIdx = 4;
  PbServerId srvId = 5;
}

message PbAppendEntriesMessage {
  uint64 term = 1;
  PbServerId leaderId = 2;
  uint64 prevLogIdx = 3;
  uint64 prevLogTerm = 4;
  repeated PbLogEntry logEntries = 5;
  uint64 commitIdx = 6;
}

message PbRequestVoteMessage {
  uint64 term = 1;
  PbServerId candidateId = 2;
  uint64 lastLogIdx = 3;
  uint64 lastLogTerm = 4;
}

message PbAddLogMessage {
  string opName = 1;
  string args = 2;
  uint64 prmIdx = 3;
  PbServerId srvId = 4;
}

message PbAddLogReply {
  bool success = 1;
  PbServerId leaderId = 2;
}

service RaftRpc {
  rpc AddLog (PbAddLogMessage) returns (PbAddLogReply) {}

  rpc AppendEntries (PbAppendEntriesMessage) returns (PbReply) {}

  rpc RequestVote (PbRequestVoteMessage) returns (PbReply) {}
}

